2 Niagara Fundamentals Exercise - Multiplexer/multiplexer-rt/src/com/tridiumuniversity/multiplexer/BMultiPlexN.java

```diff
@@ -0,0 +1,168 @@
+
+  /**
+   * Get the {@code out} property.
+   * @see #out
+   */
+  public boolean getOut() { return getBoolean(out); }
+
+  /**
+   * Set the {@code out} property.
+   * @see #out
+   */
+  public void setOut(boolean v) { setBoolean(out, v, null); }
+
+  //endregion Property "out"
+
+  //region Action "addInput"
+
+  /**
+   * Slot for the {@code addInput} action.
+   * @see #addInput()
+   */
+  public static final Action addInput = newAction(0, null);
+
+  /**
+   * Invoke the {@code addInput} action.
+   * @see #addInput
+   */
+  public void addInput() { invoke(addInput, null, null); }
+
+  //endregion Action "addInput"
+
+  //region Action "addSelector"
+
+  /**
+   * Slot for the {@code addSelector} action.
+   * @see #addSelector()
+   */
+  public static final Action addSelector = newAction(0, null);
+
+  /**
+   * Invoke the {@code addSelector} action.
+   * @see #addSelector
+   */
+  public void addSelector() { invoke(addSelector, null, null); }
+
+  //endregion Action "addSelector"
+
+  //region Type
+
+  @Override
+  public Type getType() { return TYPE; }
+  public static final Type TYPE = Sys.loadType(BMultiPlexN.class);
+
+  //endregion Type
+
+//@formatter:on
+//endregion /*+ ------------ END BAJA AUTO GENERATED CODE -------------- +*/
+
+  @Override
+  public void changed(Property p, Context cx)
+  {
+    if (!isRunning()) return;
+    calculate();
+  }
+  @Override
+  public void added(Property p, Context cx)
+  {
+    countDynamicsSlots();
+  }
+  @Override
+  public void removed(Property property, BValue oldValue, Context context)
+  {
+    countDynamicsSlots();
+  }
+  public void doAddInput()
+  {
+    //To-do maximum number of input slots
+    int flags = Flags.SUMMARY | Flags.DEFAULT_ON_CLONE;
+    add("in0?", BBoolean.make(false).as(BBoolean.class), flags); // BBoolean.make(false).as(BBoolean.class).getBoolean()
+  }
+  public void doAddSelector()
+  {
+    //To-do minimum number of input slots
+    int flags = Flags.SUMMARY | Flags.DEFAULT_ON_CLONE;
+    add("s0?", BBoolean.make(false).as(BBoolean.class), flags);
+ }
+
+
+ public void calculate()
+ {
+   String nameBuilder = "";
+   List<BBoolean> inputsList = new ArrayList<>();     //List of all Boolean inputs
+   List<BBoolean> selectorsList = new ArrayList<>();  //List of all Boolean selectors
+
+   for(int i = 0; i < countInputs; i++) {
+     nameBuilder = "in" + i;
+     BValue value = get(getSlot(nameBuilder).asProperty());
+       inputsList.add((BBoolean) value);
+   }
+
+   for(int i = 0; i < countSelectors; i++) {
+     nameBuilder = "s" + i;
+     BValue value = get(getSlot(nameBuilder).asProperty());
+     selectorsList.add(0, (BBoolean) value);
+   }
+
+   int selectedInput = 0;
+   for (int i = 0; i < countSelectors; i++) {
+     if (selectorsList.get(i).getBoolean()) {
+       selectedInput |= 1 << (countSelectors - 1 - i);
+     }
+   }
+
+   setOut(inputsList.get(selectedInput).getBoolean());
```

Comment:
This line should probably be wrapped with some error-checking to see if the number of inputs and the number of selectors makes sense. If I invoked your addSelector Action 10 times to add selectors s0 through s9 but didn't invoke addInput any, then when I change s9 to true, I think this line will produce an error because `inputsList` doesn't contain 512 elements.

Also, I don't think you need to build the `inputsList` at all. Once you've computed `selectedInput`, you could just do something like:

```java
setOut(get("in" + selectedInput));
```

Reply:
that's true, but it was intended